### PR TITLE
Add experimental `model_json_schema` function for Pydantic V2 compatibility

### PR DESCRIPTION
**Pull request by serinamarie**
As part of compatibility for Pydantic V2 in OSS, we add a `model_json_schema` function. This function aims to supplant the `cls.schema()` method prevalent in our current Pydantic v1 `BaseModel` instances.

In Pydantic V2, `model_json_schema` is a class method; but we've implemented it as a standalone function here (for now).

Note: I **do not** recommend the `unified` diff here, it is confusing.

### CODE REVIEW

File: src/prefect/_internal/pydantic/_compat.py [hunk 1]
```diff
@@ -51,19 +72,61 @@ def model_dump(
...
+
+    Returns
+    -------
+    dict[str, Any]
+        The JSON schema for the given model class.
+    """
+    if is_pydantic_v2_compatible(
+        _model_instance=model_instance, fn_name="model_json_schema"
+    ):
+        schema_generator = GenerateJsonSchema
```
Comment (serinamarie) [inline src/prefect/_internal/pydantic/_compat.py:121]: I would welcome a better way to do this

File: src/prefect/_internal/pydantic/_compat.py [hunk 2]
```diff
@@ -13,6 +13,71 @@
...
+            f"Expected a Pydantic model, but got {type(model_instance).__name__}"
+        )
+
+    should_dump_as_v2_model = (
+        HAS_PYDANTIC_V2 and PREFECT_EXPERIMENTAL_ENABLE_PYDANTIC_V2_INTERNALS
+    )
+
+    if should_dump_as_v2_model:
+        logger.debug(
+            f"Using Pydantic v2 compatibility layer for `{fn_name}`. This will be removed in a future release."
```
Comment (serinamarie) [inline src/prefect/_internal/pydantic/_compat.py:66]: Not sure how I feel about this, would welcome a thought on interpolation.

**Review by zzstoatzz — COMMENTED** [review]
this looks mostly good to me, except I think we should be representing with the types that `.model_json_schema()` (`.schema()` of old) is [typically](https://github.com/search?q=repo%3APrefectHQ%2Fprefect%20.schema()&type=code) used as (and [listed](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_json_schema) as) a `@classmethod` instead of an instance method.

TIL that it does work on an instance :)

**Comment by serinamarie** [comment]
> this looks mostly good to me, except I think we should be representing with the types, variable names, and usage in the tests that `.model_json_schema()` (`.schema()` of old) is [typically](https://github.com/search?q=repo%3APrefectHQ%2Fprefect%20.schema()&type=code) used as (and [listed](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_json_schema) as) a `@classmethod` instead of an instance method, i.e. `model_instance: BaseModel` -> `model: Type[BaseModel]`
> 
> TIL that it does work on an instance :) , but the nature of the instance should not have any bearing on the class schema

ah yeah that makes sense, haha

**Review by zzstoatzz — APPROVED** [review]
nice! i'm thinking as we expand we might want to generalize the signature of the `is_pydantic_v2_compatible` util, but imo no need to get overzealous up front

**Comment by serinamarie** [comment]
> nice! i'm thinking as we expand we might want to generalize the signature of the is_pydantic_v2_compatible util, but imo no need to get overzealous up front

I agree, I think with more commonalities between functions, we can probably remove checking the model instance there and move it somewhere else, for example

**Pull request merged**